shared/libraries/websocket_streaming/src/streaming_client.cpp

```diff
@@ -235,92 +245,149 @@ void StreamingClient::onMessage(const daq::streaming_protocol::SubscribedSignal&
     {
-        auto descriptor = sInfo.dataDescriptor;
-        onNewSignalCallback(id, sInfo);
+        auto sInfo = SignalDescriptorConverter::ToDataDescriptor(subscribedSignal);
+        if (!signalIt->second->getSignalDescriptor().assigned())
+            onDataDescriptorCallback(id, sInfo);
+        else
+            onSignalUpdatedCallback(id, sInfo);
+        signalIt->second->setDataDescriptor(sInfo.dataDescriptor);
 
-        auto inputSignal = std::make_shared<InputSignal>();
-        inputSignal->setDataDescriptor(subscribedSignal);
-        signals[id] = inputSignal;
-    }
-    else
-    {
-        onSignalUpdatedCallback(id, sInfo);
-        signals[id]->setDataDescriptor(subscribedSignal);
+        const auto tableId = subscribedSignal.tableId();
+        signalIt->second->setTableId(tableId);
+        if (auto domainDescIt = cachedDomainDescriptors.find(tableId); domainDescIt != cachedDomainDescriptors.end())
+        {
+            if (!signalIt->second->getDomainSignalDescriptor().assigned())
+                setDomainDescriptor(signalIt->first, signalIt->second, domainDescIt->second);
+        }
     }
 }
 
 void StreamingClient::setTimeSignal(const daq::streaming_protocol::SubscribedSignal& subscribedSignal)
 {
     std::string tableId = subscribedSignal.tableId();
 
-    if (signals.count(tableId) == 0)
-        return;
-
-    auto inputSignal = signals[tableId];
-    const bool assignDomainDescriptor =
-        inputSignal->getSignalDescriptor().assigned() && !inputSignal->getDomainSignalDescriptor().assigned() ? true : false;
-    inputSignal->setDomainDescriptor(subscribedSignal);
-
-    // Sets the descriptors when first connecting
-    if (assignDomainDescriptor)
+    // check if the value signal with tableId is known
+    // and the descriptors for it are already received
+    auto sInfo = SignalDescriptorConverter::ToDataDescriptor(subscribedSignal);
+    auto signalIt = std::find_if(signals.begin(),
+                                 signals.end(),
+                                 [tableId](const std::pair<std::string, InputSignalPtr>& pair)
+                                 {
+                                     return tableId == pair.second->getTableId();
+                                 });
+    if (signalIt != signals.end())
     {
-        auto domainDescriptor = inputSignal->getDomainSignalDescriptor();
-        onDomainDescriptorCallback(tableId, domainDescriptor);
-
-        if (auto iterator = signalInitializedStatus.find(tableId); iterator != signalInitializedStatus.end())
-        {
-            iterator->second.first.set_value();
-        }
+        // value signal with tableId is known, set domain descriptor for it
+        setDomainDescriptor(signalIt->first, signalIt->second, sInfo.dataDescriptor);
+    }
+    else
+    {
+        // value signal with tableId is unknown, save domain descriptor
+        cachedDomainDescriptors.insert_or_assign(tableId, sInfo.dataDescriptor);
```

Comment:
Does this assume we only ever have 2 signals with the same table ID, and the signal is actually our data signal? Though I think we'll have issues with such cases in general, so this should be a good enough solution for now.

It might be better to look for a signal that has a time relation to the domain signal we received here.

Reply:
It assumes that at least all data signals are published as `available`. There might be more than 2 signals sharing the same `tableId`: the data signal itself, related domain signal, related status signal etc. Signals other than the domain or the data signal itself will simply be ignored at the moment (in case of not being published as `available` ). 
An alternative to utilizing the `tableId` is to search for a signal with the type 'domain' within the `relatedSignals` list to identify which signal serves as the domain for the data signal. However, I don't particularly see any advantages to this approach.